gwcs/wcs.py

```diff
@@ -634,3 +641,273 @@ def fix_inputs(self, fixed):
+        Ud = ( cdmat[1][1] * undist_xd - cdmat[0][1] * undist_yd) / detd
+        Vd = (-cdmat[1][0] * undist_xd + cdmat[0][0] * undist_yd) / detd
+  
+        if max_inv_pix_error:
+            fit_inv_poly_u, fit_inv_poly_v, max_inv_resid = _fit_2D_poly(ntransform,
+                                                            npoints, None,
+                                                            max_inv_pix_error,
+                                                            U, V, u-U, v-V,
+                                                            Ud, Vd, ud-Ud, vd-Vd,
+                                                            verbose=verbose)
+        pdegree = fit_poly_x.degree
+        if pdegree > 1:
+            hdr['a_order'] = pdegree
+            hdr['b_order'] = pdegree
+            _store_2D_coefficients(hdr, sip_poly_x, 'A')
+            _store_2D_coefficients(hdr, sip_poly_y, 'B')
+            if max_inv_pix_error:
+                _store_2D_coefficients(hdr, fit_inv_poly_u, 'AP', keeplinear=True)
+                _store_2D_coefficients(hdr, fit_inv_poly_v, 'BP', keeplinear=True)
+            hdr['sipmxerr'] = (max_resid * plate_scale, 'Max diff from GWCS (equiv pix).')
+            if max_inv_pix_error:
+                ipdegree = fit_inv_poly_u.degree
+                hdr['ap_order'] = ipdegree
+                hdr['bp_order'] = ipdegree
+                hdr['sipiverr'] = (max_inv_resid, 'Max diff for inverse (pixels)')
+        else:
+            hdr['ctype1'] = 'RA---TAN'
+            hdr['ctype2'] = 'DEC--TAN'
+
+        hdr['cd1_1'] = cdmat[0][0]
+        hdr['cd1_2'] = cdmat[0][1]
+        hdr['cd2_1'] = cdmat[1][0]
+        hdr['cd2_2'] = cdmat[1][1]
+        return hdr
+
+
+def _fit_2D_poly(ntransform, npoints, degree, max_error,
+                 xin, yin, xout, yout,
+                 xind, yind, xoutd, youtd,
+                 verbose=False):
+    """
+    Fit a pair of ordinary 2D polynomial to the supplied transform.
+
+    Use adaptive sampling based on the current degree.
+    """
+    llsqfitter = LinearLSQFitter()
+
+    # The case of one pass with the specified polynomial degree
+    if degree:
+        deglist = [degree]
+    else:
+        deglist = range(10)
+    prev_max_error = -1
+    if verbose:
+        print(f'maximum_specified_error: {max_error}')
+    for deg in deglist:
+        poly_x = Polynomial2D(degree=deg)
+        poly_y = Polynomial2D(degree=deg)
+        fit_poly_x = llsqfitter(poly_x, xin, yin, xout)
+        fit_poly_y = llsqfitter(poly_y, xin, yin, yout)
+        max_resid = _compute_distance_residual(xout, yout,
+                                               fit_poly_x(xin, yin),
+                                               fit_poly_y(xin, yin))
+        if prev_max_error < 0:
+            prev_max_error = max_resid
+        else:
+            if max_resid > prev_max_error:
+                raise RuntimeError('Failed to achieve required error tolerance')
```

Comment:
As a consequence of https://github.com/spacetelescope/gwcs/pull/286/files#r390430351, this code can now be simplified to:
```suggestion
        if max_resid > prev_max_error:
            raise RuntimeError('Failed to achieve required error tolerance')
        prev_max_error = max_resid
```

Reply:
`prev_max_error = max_resid` is a little bit hard to "read": sometimes you call errors `resid`(uals) and sometimes `error` in variable names. It would be nicer for maintainability to be consistent with naming.

Reply:
I really do not think that a `RuntimeError` should be raised at all. I think the code should stop iterating and return the best known result (and if you insist, print a warning). When the error does not increase as the degree is increased **but** the requested accuracy is not achieved, current code will simply return without any exception and without printing the `'terminating condition met'` message. That all. Why is the case when errors increase any different from the previously described scenario.

Reply:
Done. As for the runtime error, I think we should raise. Make the user very aware that the condition isn't met. I'm sure many may not notice, particularly if it used inside code.

Reply:
I disagree with regard to the `RuntimeError`.

Also, you can simply press the "Commit suggestion" button if you agree with the proposed code.